js/src/common/Application.tsx

```diff
@@ -24,18 +24,80 @@ import PageState from './states/PageState';
 import ModalManagerState from './states/ModalManagerState';
 import AlertManagerState from './states/AlertManagerState';
 
+import type DefaultResolver from './resolvers/DefaultResolver';
+import type Mithril from 'mithril';
+import type Component from './Component';
+import { ComponentAttrs } from './Component';
+
+/**
+ * A valid route definition.
+ */
+export type RouteItem<
+  Attrs extends ComponentAttrs,
+  Comp extends Component<Attrs & { routeName: string }>,
+  RouteArgs extends Record<string, unknown> = {}
+> = {
+  /**
+   * The path for your route.
+   *
+   * This might be a specific URL path (e.g.,`/myPage`), or it might
+   * contain a variable used by a resolver (e.g., `/myPage/:id`).
+   *
+   * @see https://docs.flarum.org/extend/frontend-pages.html#route-resolvers-advanced
+   */
+  path: `/${string}`;
+} & (
+  | {
+      /**
+       * The component to render when this route matches.
+       */
+      component: { new (): Comp };
+      /**
+       * A custom resolver class.
+       *
+       * This should be the class itself, and **not** an instance of the
+       * class.
+       */
+      resolverClass?: { new (): DefaultResolver<Attrs, Comp, RouteArgs> };
+    }
+  | {
+      /**
+       * An instance of a route resolver.
+       */
+      resolver: RouteResolver<Attrs, Comp, RouteArgs>;
+    }
+);
+
+export interface RouteResolver<
```

Comment:
Why can't we use Mithril's `RouteResolver`?

Reply:
This is more specific to our use case, in terms of typing attributes and the fact that Flarum uses its Component class for basically every component.